counter/src/main/java/org/infinispan/counter/impl/CounterModuleLifecycle.java

```diff
@@ -102,6 +103,10 @@ private static void registerCounterManager(GlobalComponentRegistry registry, Com
             counterManager = new EmbeddedCounterManager(future,
                   registry.getGlobalConfiguration().globalState().enabled());
             registry.registerComponent(counterManager, CounterManager.class);
+            //this start() is only invoked when the DefaultCacheManager.start() is invoked
+            //it is invoked here again to force it to check the managed global components
+            // and register them in the MBeanServer, if they are missing.
+            registry.getComponent(CacheManagerJmxRegistration.class).start(); //HACK!
```

Comment:
WDYT about removing the automation and allowing each component to register on `start()` instead?

Reply:
@danberindei I don't follow, could you elaborate?

Reply:
@rvansa the component registry allows components to be created lazily, and these lazy components can't be exposed in JMX by `CacheJmxRegistration.start()` because they don't exist yet. So I was thinking of letting each MBean component call something like `CacheManagerJmxRegistration.registerMBean(this)` in their own `start()` method.

There are other options of course, e.g. we could also add listeners to the component registry for `CaheManagerJmxRegistration` to automatically register new components in JMX, or we could make sure `CacheManagerJmxRegistration.start()` runs last and throw an exception in `AbstractComponentRegistry` if a MBean component is added after`AbstractComponentRegistry.start()`.

Reply:
-1 to change the automated process, however...
+1 to add a `registerMbean` method! 

AFAIK, all the `core` component are registered before `CacheJmxRegistration.start())`, but in the remaining modules (counters, query, etc), it happens after the `CacheJmxRegistration.start()`. So, your last sentence won't work.

Reply:
@pruivo it could be that all the core components are registered, but I know for sure there isn't any check: the transport used to be started only with the first cache. 

For the other modules, I don't think `DefaultCacheManager` *has* to call `CacheJmxRegistration.start())` before all the modules are initialized, it's just how it used to work. Based on the ISPN-118 commit message I'd say stoppping the global component registry shouldn't deregister the cache manager MBean, but it should deregister all the other MBeans.